### Introduction

This Folder contains the tools and methods developed for WP2 for the course
“Applied data analysis in bioinformatics” from the masters program “Bioinformatik und Systembiologie”
at the Justus-Liebig-University and the Technische Hochschule Mittelhessen in the winter term 2022/2023.
</br>The goal of this course is to develop a pipeline for the [Max Planck Institute for Heart and Lung Research](https://www.mpg.de/149809/heart-lung-research)
which takes data from [CATLAS](http://catlas.org/humanenhancer/#!/) and performs distinct analyses mainly based on the
chromatin accessibility for different human tissues<sup>[1](#--1-zhang-k-hocker-j-d-miller-m-hou-x-chiou-j-poirion-o-b-qiu-y-li-y-e-gaulton-k-j-wang-a-preissl-s-amp-ren-b--2021---a-single-cell-atlas-of-chromatin-accessibility-in-the-human-genome-cell-184--24---httpsdoiorg101016jcell202110024)</sup>.
Furthermore, this pipeline is organized into two separate packages (WP1/WP2) due to the group distribution in the course.

This WP worked on tools that generates and adds new data to the already existing
[Anndata](https://anndata.readthedocs.io/en/latest/)<sup>[2](#font-size1---2-httpsscanpyreadthedocsioenstable-font)</sup>
Object from [CATLAS](http://catlas.org/humanenhancer/#!/) and visualizing them.
The data generated by WP2 focuses mainly
on the Overlap  of Reads to a given feature mapped on a reference genome. Such a feature
can be for example the promoter-region or the enhancer-region of a gene.
The data can then be
visualized and checked for correlations between the generated Feature data via Violin, UMAP and Scatter Plots.
Furthermore, the data created by WP1 can also be visualized with these tools.


In the following we will talk about how to acquire the data we need and how to find or produce the informations we want.
In the Chapter [Quick Start](#quick-start) you'll find an Introduction on how to use the methods.


1. [The data source](#The-data-source)
   1. [Calculating the Data](#Calculating-the-Data)
2. [Reading a General Feature File](#Reading-a-General-Feature-File)
   1. [Annotated Features](#Annotated-Features)
   2. [Other Features](#Other-Features)
3. [Calculating Feature Overlap](#Calculating-Feature-Overlap)
4. [Data Presentation](#Data-Presentation)
   1. [Different Presentation Styles](#Different-Presentation-Styles)
6. [Functions](#Functions)
   1. [Dimension Reduction](#Dimension-Reduction)
   2. [Feature Comparison in Violinplots](#Feature-Comparison-in-Violinplots)
7. [References](#References)
8. [Quick Start](#quick-start)
9. [Authors](#Authors)

### The data source

For the purpose of this course, we used the Human reference genome hg38 accessible from [NCBI](https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.26/)<sup>[3](#font-size1---3-httpswwwncbinlmnihgovassemblygcf00000140526-font)</sup>.
This reference Genomes comes in the [General-Feature-File](https://www.ensembl.org/info/website/upload/gff.html)(GFF)<sup>[4](#font-size1---4-httpswwwensemblorginfowebsiteuploadgffhtml-font)</sup>
Format, which is a standardized way of storing informations about regions in a Genome inside a text
file. This text file consists of a possible Header entry and a line for every region in the genome, with each line having
9 columns of data entrys.
Below you can find an example for a [GFF](https://www.ensembl.org/info/website/upload/gff.html) file from the
[GMOD](http://gmod.org/wiki/GFF3)<sup>[5](#font-size1---5-httpgmodorgwikigff3-font)</sup> Wiki.
<p align="center">
   <img src="docs/images/gff_file_excerpt.png" />
</p>

The next table provides information about each column a line can have, empty columns are denoted with a '.' <sup>[3](#font-size1---3-httpswwwensemblorginfowebsiteuploadgffhtml-font)</sup>.

| Column Number 	| Name      	| Description                                                                                                                                                         	|
|---------------	|-----------	|---------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| 1             	| seqname   	| Name of the chromosome or scaffold                                                                                                                                  	|
| 2             	| source    	| Programm that generated the feature, or the source of the data                                                                                                      	|
| 3             	| feature   	| Feature type                                                                                                                                                        	|
| 4             	| start     	| Start position of the feature on the reference genome                                                                                                               	|
| 5             	| end       	| End position of the feature on the reference genome                                                                                                                 	|
| 6             	| score     	| Floating point value                                                                                                                                                	|
| 7             	| strand    	| + (forward) or - (reverse)                                                                                                                                          	|
| 8             	| frame     	| One of '0', '1' or '2'. '0' indicates that the first base of the feature<br>is the first base of a codon, '1' that the second base is the first base <br>of a codon 	|
| 9             	| attribute 	| List of semicolon-seperated additional features                                                                                                                     	|

Also this WP uses the same fragment files provided by [CATLAS](http://catlas.org/humanenhancer/#!/) as used by WP1,
with the same information needed.
A brief description of the `.bed` file Format the fragment file comes in can be found in their respective [README.md](/datenanalyse/WP1/README.md)
`The data source` Chapter. </br>
At last we need a `.h5ad` File, correpsonding to the fragment file we use. This file
contains an prepared [Anndata](https://anndata.readthedocs.io/en/latest/) object from [CATLAS](http://catlas.org/humanenhancer/#!/)
where the later on generated data is stored to.

#### Calculating the Data

To calculate a feature overlap, we first need to know where the specific features are
located on the reference genome. These features are annotated in the third column of the [GFF](https://www.ensembl.org/info/website/upload/gff.html) file.
This column holds information about the feature type of the specific line, with a predefined vocabulary <sup>[6](#font-size1---6-httpswwwinsdcorgsubmitting-standardsfeature-table-font)</sup>.
The fourth and fifth Column in this line describes the Start and Stop Position of this feature on the reference
genome. For later calculations it is mandatory to generate separate [GFF](https://www.ensembl.org/info/website/upload/gff.html) files,
where each file containing only lines with the same feature. We talk on how to use the provided tools for
generating these files in the next Chapter.</br>
The location of a read inside a [fragment](/datenanalyse/WP1/README.md) file, can be found in the second and third column of a line.
By counting the number of reads from the fragment file that's position is inside the location of a feature
from the reference genome we can finally calculate the feature overlap rate.


### Reading a General Feature File

To generate the feature filtered [GFF](https://www.ensembl.org/info/website/upload/gff.html) file, the package
_gff\_analyser_ provides the two modules:

- gffBuilder.py
- gffClasses.py

The _gffBuilder.py_ contains a parser for the [GFF](https://www.ensembl.org/info/website/upload/gff.html) file format,
which stores the file information into an object and a wrapper function to generate all possible feature files. The parser is used as follows:

      object_list = build_gff3_class(file=gtf_file)
Inside the list _object\_list_ one object for each input file is stored. By iterating with a `for`-loop
over this list we can now use the methods of these objects. It would be possible here to replace the
variable `features` with a a list containing only 1 string of the desired feature to investigate, so that only a file of this specific
feature is generated.

The _gffClasses.py_ contains all the methods needed for generating the [GFF](https://www.ensembl.org/info/website/upload/gff.html)
files. A more detailed use of these methods can be found in following Chapters.

#### Annotated Features

 To generate a file for each feature annotated inside
these objects, we can use the method `self.generate_feature_gtf(gffdata_list=object_list, feature_keys=features)`
</br>The input the method uses are:
- gffdata_list = Previously generated object_list
- feature_keys = A list of features to analyse

Below an example for generating all possible feature files is shown:

       for element in organism_list:
            features = element.count_features()        
            element.generate_feature_gtf(gffdata_list=object_list, feature_keys=features)

The method `features = element.count_features()` returns a list of all existing features inside the
[GFF](https://www.ensembl.org/info/website/upload/gff.html) file, so the method `element.generate_feature_gtf(feature_keys=features)`
knows what files exactly to generate, as this list is used as positional argument in this method.
The generated files will now be placed inside an _out/_ folder, created inside the current working directory.

#### Other Features

Some features like promoters are not directly annotated inside a [GFF](https://www.ensembl.org/info/website/upload/gff.html) file,
but can be calculated by the gene model as shown exemplarily in the image<sup>[7](#font-size1---7-httpswwwresearchgatenetfigurean-illustration-of-a-typical-structure-of-a-eukaryotic-gene-a-gene-may-have-manyfig351510353-font)</sup> below:

<p align="center">
   <img src="docs/images/An-illustration-of-a-typical-structure-of-a-eukaryotic-gene-A-gene-may-have-many.png" />
</p>

The features this WP focused on are as listed:

- Promoter
- Transcription Start Site (TSS)
- Gene Body
- Peaks
- Blacklisted Regions
- Enhancer

The information for the enhancers and blacklisted regions are accessed thorugh
the database [NCBI](https://www.encodeproject.org/references/ENCSR938RZZ/), where
a `.bed` file with the sequence regions is provided.</br>
With the information gathered in the [previous Chapter](#Annotated-Features)
and the database, we can now calculate the positions of these new features.
The methods provided in _gffClasses.py_ either re-evaluate the start and end Positions of
the corresponding objects inside the _object\_list_ or make use of _Bedtools_<sup>[8](#font-size1---8-httpsbedtoolsreadthedocsioenlatest-font)</sup> 
and write them to a file inside the output directory.

##### Promoter

To get a promoters start and stop Position, we first need to know where a gene is located. From the genes start 
position we can subtract a stated length which represents the corresponding promotors start position. The end position
of this promotor is the start of the corresponding gene.</br>
To generate a [GFF](https://www.ensembl.org/info/website/upload/gff.html) file containing only promoter regions 
we can call the method: 
            
      self.generate_promoter_gtf(gffdata_list: list, promoter_distance: int, out='out/')

The input needed for this method is:

- gffdata_list = Previously generated object_list
- promoter_distance = Stated length of promoter, for example `2000`
- out = Output Directory, by default `out/`

##### TSS

To get the TSS start and stop Position, we also need to know where a gene is located. From the gene start, we
can add a stated length to the start to get the end position of the TSS.</br> The method for generating a TSS feature file is:

      element.generate_tss_gtf(gffdata_list: list, tss_distance: int, out=out/)

The input needed for this method is:

- gffdata_list = Previously generated object_list
- TSS_distance = Stated length of the TSS, for example `100`
- out = Output Directory, by default `out/`

##### Gene Body

The gene bodies can be found by subtracting the Five and Three prime UTR from the gene feature annotated
lines in the [GFF](https://www.ensembl.org/info/website/upload/gff.html) file. To create this file, a Version of
a [GFF](https://www.ensembl.org/info/website/upload/gff.html) file containing only genes must be previously generated 
and placed inside the output directory for this method to work.</br>
The method for generating a Gene Body feature file is:

      element.generate_gene_body_gtf(gffdata_list=object_list, out=out)

The input needed for this method is:

- gffdata_list = Previously generated object_list
- out = Output Directory, by default `out/`

##### Peaks

---> Implement the features into gff analyser!!

##### Blacklisted Regions

---> Implement the features into gff analyser!!

##### Enhancer

---> Implement the features into gff analyser!!

### Calculating Feature Overlap

---> sc toolbox wrapper function

### Data Presentation

After preparing the data for presentation with some normalization and filtering, such as, but not limited to, removing chromosomes X, Y (Optional) and M, removing cells without features or empty features.
These functions are combined into the `presentation.py` script, to avoid unnecessary clutter in any `Jupyter Notebook` you might use them in.

#### Different Presentation Styles

- Violinplots
  - display density accross celltypes and featurerate
- Correlation Plots
  - scatterplots that compare two metrics
  - can visualize linear or a parabolical correlation
- Dimension Reduction Maps (umaps)

### Functions

#### Dimension Reduction
```py
compareDimesionreductions(adata, key: str or list, comparator: str or list)
```

This function displays the `Dimension Reductions` defined in key and lastly add the `Comperator` or `Comperators` to compare the prior maps to. The comparator will be visible in each row while the dimension reductions labelled by `key` will be one per row.

#### Feature Comparison in Violinplots

```py
def compareFeatureToCelltypes(adata, feature: list or str, comparator: str, save=None):
```

The concept is similar to the previous mentioned function, but it focuses on `violin plots`. It expects either a single or a list of strings that represent a feature and a comparator that is used to group these together, in practice we used the predicted celltypes.




### Quick Start


### References



#### <font size=1>- [1] Zhang, K., Hocker, J. D., Miller, M., Hou, X., Chiou, J., Poirion, O. B., Qiu, Y., Li, Y. E., Gaulton, K. J., Wang, A., Preissl, S., &amp; Ren, B. (2021). A single-cell atlas of chromatin accessibility in the human genome. Cell, 184(24). https://doi.org/10.1016/j.cell.2021.10.024 </font>
#### <font size=1>- [2] https://scanpy.readthedocs.io/en/stable/ </font>
#### <font size=1>- [3] https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.26/ </font>
#### <font size=1>- [4] https://www.ensembl.org/info/website/upload/gff.html </font>
#### <font size=1>- [5] http://gmod.org/wiki/GFF3 </font>
#### <font size=1>- [6] https://www.insdc.org/submitting-standards/feature-table/ </font>
#### <font size=1>- [7] https://www.researchgate.net/figure/An-illustration-of-a-typical-structure-of-a-eukaryotic-gene-A-gene-may-have-many_fig3_51510353 </font>
#### <font size=1>- [8] https://bedtools.readthedocs.io/en/latest/ </font>




### Authors

    Daniel Tischler
        daniel.tischler@bioinfsys.uni-giessen.de

    Noah Leon Stürtz
         noah.leon.stuertz@bioinfsys.uni-giessen.de
---

### Appendix
